Fix duplicated path.Contains check and other checks in Path.GetFullPath

steveberdy opened the pull request:
## Summary

Refactors code to prevent duplicate checks on paths in the `Path.GetFullPath` methods.

Fixes #54993

steveberdy committed:
Update Path.Windows.cs

steveberdy committed:
Merge branch 'dotnet:main' into main

iSazonov commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:
```diff
@@ -49,17 +49,8 @@ public static string GetFullPath(string path)
             // unpredictable results.
             if (path.Contains('\0'))
```
Initial issue is about this line.
The simplest fix would be splitting  Path.GetFullPath() on public (contains input checks) and internal (contains rest of the code).

steveberdy replied:
Ok. I'll switch the private to internal and make sure to remove any trailing whitespace on the two lines the checks pointed out. Thanks!

iSazonov replied:
I'm sorry, I mean there's a lot of GetFullPath() usings in the libraries. Of course you can stop on the initial fix.
For example, see GetVersionInfo(), methods in class Directory/File and many, many others where we could use the internal GetFullQualifiedPath().

steveberdy replied:
Ok. I'll submit a separate PR today, since these tests are almost done. If they approve the new refactoring, I'll apply it throughout the rest of the script.

steveberdy committed:
Change helper method to internal

Switched method from a private protection level to an internal protection level. Also removed trailing whitespace.

steveberdy commented:
@adamsitnik @carlossanlop @jozkee Please review this refactoring. If you approve, I'll add another commit to refactor the rest of the `Path.Windows.cs` script. Thanks.

steveberdy committed:
Merge branch 'dotnet:main' into main

steveberdy committed:
Merge branch 'dotnet:main' into main

steveberdy commented:
@Jozkee please review

steveberdy committed:
Merge branch 'dotnet:main' into main

steveberdy committed:
Merge branch 'dotnet:main' into main

steveberdy commented:
@stephentoub Please review

steveberdy committed:
Merge branch 'dotnet:main' into main

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:
```diff
@@ -50,16 +50,7 @@ public static string GetFullPath(string path)
-            }
-
-            return PathHelper.Normalize(path);
+            return GetFullQualifiedPath(path);
```
```suggestion
            return GetFullyQualifiedPath(path);
```

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:
```diff
@@ -77,7 +68,7 @@ public static string GetFullPath(string path, string basePath)
 
             if (IsPathFullyQualified(path))
-                return GetFullPath(path);
+                return GetFullQualifiedPath(path);
```
```suggestion
                return GetFullyQualifiedPath(path);
```

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:
```diff
@@ -129,7 +120,21 @@ public static string GetFullPath(string path, string basePath)
             return PathInternal.IsDevice(combinedPath.AsSpan())
                 ? PathInternal.RemoveRelativeSegments(combinedPath, PathInternal.GetRootLength(combinedPath.AsSpan()))
-                : GetFullPath(combinedPath);
+                : GetFullQualifiedPath(combinedPath);
```
```suggestion
                : GetFullyQualifiedPath(combinedPath);
```

jeffhandley commented on src/libraries/System.Private.CoreLib/src/System/IO/Path.Windows.cs:
```diff
@@ -129,7 +120,21 @@ public static string GetFullPath(string path, string basePath)
+                : GetFullQualifiedPath(combinedPath);
+        }
+
+        internal static string GetFullQualifiedPath(string path)
```
```suggestion
        internal static string GetFullyQualifiedPath(string path)
```

jeffhandley reviewed (approved):
Nit suggestion on renaming the method. I'll go ahead and commit the suggestions and merge once checks rerun/pass.

jeffhandley committed:
Rename method to GetFullyQualifiedPath per code review

steveberdy commented:
Thank you @jeffhandley!
